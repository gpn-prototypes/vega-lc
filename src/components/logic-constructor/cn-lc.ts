import bemCn from 'bem-cn';

export const cnLogicConstructor = bemCn('LogicConstructorWidget');
