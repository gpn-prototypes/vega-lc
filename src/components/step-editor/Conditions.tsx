import React from 'react';

const Conditions: React.FC = () => {
  return <div>Conditions</div>;
};

export default Conditions;
