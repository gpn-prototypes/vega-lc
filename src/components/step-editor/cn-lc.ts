import bemCn from 'bem-cn';

export const cnStepEditor = bemCn('StepEditor');
