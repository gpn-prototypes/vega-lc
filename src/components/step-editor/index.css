:root {
  --step-editor-content-height: calc(100% - 50px);
  --step-editor-select-top-indent: 12px;
}

.StepEditor {
  position: fixed;
  top: 70px;
  right: 30px;

  width: 528px;
  height: calc(100vh - 100px);

  &__Tabs {
    position: relative;

    margin-bottom: 2px;
    padding-left: 12px;

    height: 35px;

    background-color: var(--color-bg-default);

    & > .Tabs {
      position: absolute;

      top: 2px;
    }
  }

  &__Content {
    padding: 24px var(--step-editor-select-top-indent);

    width: 100%;
    height: var(--step-editor-content-height);

    background-color: var(--color-bg-default);
  }

  &__ChanceSelect {
    width: 100%;
  }

  &__SelectLabel {
    margin: 0 0 8px;

    display: flex;
    align-items: center;

    font-size: 14px;
    line-height: 120%;

    color: var(--color-typo-secondary);
  }

  &__TimeFrames {
    margin-top: var(--step-editor-select-top-indent);

    width: 100%;
  }

  &__ExtremePoints {
    display: flex;

    width: 100%;

    & > div {
      width: 50%;
    }

    & > div:first-of-type {
      margin-right: 12px;
    }
  }

  &__DatePointSelect {
    width: 100%;
  }

  &__Continuous {
    margin-top: var(--step-editor-select-top-indent);
  }

  &__ContinuousSelect {
    width: calc(50% - 6px);
  }

  &__Cost {
    margin-top: var(--step-editor-select-top-indent);
  }

  &__CostSelect {
    width: calc(50% - 6px);
  }

  &__ParameterAppender {
    margin-top: calc(var(--step-editor-select-top-indent) * 2);
  }
}
