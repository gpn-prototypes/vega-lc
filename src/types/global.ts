import ResizeObserver from 'resize-observer-polyfill';

declare global {
  interface Window {
    ResizeObserver: typeof ResizeObserver;
  }
}
