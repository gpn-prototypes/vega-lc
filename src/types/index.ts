export interface Identity {
  getToken(): Promise<string>;
}
