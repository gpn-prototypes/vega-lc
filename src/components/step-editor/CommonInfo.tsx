import React from 'react';
import { BasicSelect, Button, IconAdd, IconTrash } from '@gpn-prototypes/vega-ui';

import { cnStepEditor } from './cn-lc';

type Option = {
  label: string;
  value: string;
};

type Value = {
  label: string;
  value: unknown;
};

type Item = {
  value: Value;
  id: string;
  section: 'Parameters' | 'Chances';
};

const CommonInfo: React.FC = () => {
  const chanceOptions: Value[] = [
    { label: 'Обстановка осадконакопления, Морская', value: '1' },
    { label: 'Водянистая почва', value: '2' },
  ];

  const costOptions: Value[] = [
    { label: '1 000 000 рублей', value: '1' },
    { label: '5 000 000 рублей', value: '2' },
    { label: '10 000 000 рублей', value: '3' },
  ];

  const deleteItem = (id: string): void => {
    console.log(id);
  };

  const getItemLabel = (option: Option): string => option.label;

  const chanceItems: Item[] = [
    {
      id: '23ff9b35-d2f1-4af1-a766-9e6c14e162f8',
      value: { label: 'Обстановка осадконакопления, Морская', value: '1' },
      section: 'Chances',
    },
    {
      id: 'd7590f12-4ec5-45ac-859a-0c6a95919899',
      value: { label: 'Водянистая почва', value: '2' },
      section: 'Chances',
    },
  ];

  const costItems: Item[] = [
    {
      id: '9d9d871b-9355-4515-85e1-70cfe405f8e5',
      value: { label: '10 000 000 рублей', value: '3' },
      section: 'Parameters',
    },
  ];

  const renderItems = (items: Item[]) => {
    return items.map((item) => {
      return (
        <div key={item.id} className={cnStepEditor('SelectContainer')}>
          <BasicSelect
            size="s"
            className={cnStepEditor(`${item.section}Select`).toString()}
            value={item.value}
            options={item.section === 'Chances' ? chanceOptions : costOptions}
            id={item.id}
            getOptionLabel={getItemLabel}
            view="default"
          />

          {item.section === 'Chances' && (
            <IconTrash
              onClick={() => deleteItem(item.id)}
              className={cnStepEditor('RemoveBtn').toString()}
              size="xs"
              view="secondary"
            />
          )}
        </div>
      );
    });
  };

  return (
    <div className={cnStepEditor('CommonInfo')}>
      <div className={cnStepEditor('Chances')}>
        <p className={cnStepEditor('SectionLabel').toString()}>Риск</p>

        {renderItems(chanceItems)}

        <div className={cnStepEditor('ItemAppender')}>
          <Button
            width="default"
            type="button"
            view="ghost"
            size="s"
            iconSize="xs"
            iconLeft={IconAdd}
            onClick={(): void => {}}
            label="Добавить риск"
          />
        </div>
      </div>

      <div className={cnStepEditor('Parameters')}>
        <p className={cnStepEditor('SectionLabel').toString()}>Параметры</p>

        <p className={cnStepEditor('SelectLabel').toString()}>Стоимость</p>

        {renderItems(costItems)}

        <div className={cnStepEditor('ItemAppender')}>
          <Button
            width="default"
            type="button"
            view="ghost"
            size="s"
            iconSize="xs"
            iconLeft={IconAdd}
            onClick={(): void => {}}
            label="Добавить параметр"
          />
        </div>
      </div>
    </div>
  );
};

export default CommonInfo;
