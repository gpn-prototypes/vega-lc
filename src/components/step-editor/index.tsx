import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { IconClose, Tabs } from '@gpn-prototypes/vega-ui';

import { cnStepEditor } from './cn-lc';
import CommonInfo from './CommonInfo';
import Conditions from './Conditions';
import Effects from './Effects';

import './index.css';

import { toggleStepEditor } from '@/redux-store/logic-constructor/actions';

type Item = {
  name: string;
};

const StepEditor: React.FC = () => {
  const dispatch = useDispatch();

  const handleClose = (): void => {
    dispatch(toggleStepEditor(false));
  };

  const tabs = [
    {
      name: 'Общая информация',
    },
    {
      name: 'Эффекты',
    },
    {
      name: 'Условия выполнения',
    },
  ];

  const [val, setValue] = useState<Item | null>(tabs[0]);

  const renderContent = (): React.ReactElement => {
    switch (val?.name) {
      case tabs[0].name:
        return <CommonInfo />;
      case tabs[1].name:
        return <Effects />;
      case tabs[2].name:
        return <Conditions />;
      default:
        return <CommonInfo />;
    }
  };

  return (
    <div className={cnStepEditor()}>
      <div className={cnStepEditor('Tabs')}>
        <Tabs
          size="s"
          view="clear"
          items={tabs}
          value={val}
          getLabel={(item): string => item.name}
          onChange={({ value }): void => setValue(value)}
        />
      </div>

      <div className={cnStepEditor('Content')}>
        {renderContent()}

        <IconClose
          size="xs"
          className={cnStepEditor('CloseBtn').toString()}
          onClick={handleClose}
        />
      </div>
    </div>
  );
};

export default StepEditor;
