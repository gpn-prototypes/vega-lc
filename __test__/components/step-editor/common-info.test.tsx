import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render } from '@testing-library/react';
import ResizeObserver from 'resize-observer-polyfill';

import '../../../src/types/global';

import { cnStepEditor } from '../../../src/components/step-editor/cn-lc';
import CommonInfo from '../../../src/components/step-editor/CommonInfo';

const renderComponent = () => render(<CommonInfo />);

describe('Common Info', () => {
  test('рендерится без ошибок', () => {
    window.ResizeObserver = ResizeObserver;

    const dom = renderer.create(<CommonInfo />).toJSON();
    expect(dom).toMatchSnapshot();
  });

  test('удаляется элемент', async () => {
    const component = await renderComponent();

    const removeButton = component.container.querySelector(`.${cnStepEditor('RemoveBtn')}`);

    if (removeButton) {
      fireEvent.click(removeButton);
    }

    // TODO fix after implementing onClick event
  });

  test('срабатывает кнопка "Добавить риск"', async () => {
    const component = await renderComponent();

    const addRisk = component.container.querySelector(
      `.${cnStepEditor('Chances')} > .${cnStepEditor('ItemAppender')} > button`,
    );

    if (addRisk) {
      fireEvent.click(addRisk);
    }

    // TODO fix after implementing onClick event
  });

  test('срабатывает кнопка "Добавить параметр"', async () => {
    const component = await renderComponent();

    const addParameter = component.container.querySelector(
      `.${cnStepEditor('Parameters')} > .${cnStepEditor('ItemAppender')} > button`,
    );

    if (addParameter) {
      fireEvent.click(addParameter);
    }

    // TODO fix after implementing onClick event
  });
});
