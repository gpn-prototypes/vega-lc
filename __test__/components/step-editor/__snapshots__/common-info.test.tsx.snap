// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common Info рендерится без ошибок 1`] = `
<div
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <p
      className="StepEditor__SectionLabel"
    >
      Риск
    </p>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <div
        className="Select Select_size_s Select_form_default Select_view_default StepEditor__ChancesSelect"
      >
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          className="Select-Control"
          id="23ff9b35-d2f1-4af1-a766-9e6c14e162f8"
        >
          <div
            className="Select-ControlInner"
          >
            <div
              className="Select-ControlValueContainer"
            >
              <input
                className="Select-FakeField"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                type="button"
              />
              <span
                className="Select-ControlValue"
                title="Обстановка осадконакопления, Морская"
              >
                Обстановка осадконакопления, Морская
              </span>
            </div>
          </div>
          <span
            className="Select-Indicators"
          >
            <button
              className="Select-IndicatorsDropdown"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="Icon Icon_size_xs IconSelect Select-DropdownIndicatorIcon"
              >
                <svg
                  className="Icon-Svg"
                  viewBox="0 0 12 12"
                >
                  <path
                    d="M6 8.5L10 4H2l4 4.5z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </div>
      <span
        className="Icon Icon_size_xs Icon_view_secondary IconTrash StepEditor__RemoveBtn"
        onClick={[Function]}
      >
        <svg
          className="Icon-Svg"
          viewBox="0 0 12 12"
        >
          <path
            clipRule="evenodd"
            d="M4 2a1 1 0 011-1h2a1 1 0 011 1h3v1h-1v7a1 1 0 01-1 1H3a1 1 0 01-1-1V3H1V2h3zm1 2H4v5h1V4zm3 0H7v5h1V4z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <div
        className="Select Select_size_s Select_form_default Select_view_default StepEditor__ChancesSelect"
      >
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          className="Select-Control"
          id="d7590f12-4ec5-45ac-859a-0c6a95919899"
        >
          <div
            className="Select-ControlInner"
          >
            <div
              className="Select-ControlValueContainer"
            >
              <input
                className="Select-FakeField"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                type="button"
              />
              <span
                className="Select-ControlValue"
                title="Водянистая почва"
              >
                Водянистая почва
              </span>
            </div>
          </div>
          <span
            className="Select-Indicators"
          >
            <button
              className="Select-IndicatorsDropdown"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="Icon Icon_size_xs IconSelect Select-DropdownIndicatorIcon"
              >
                <svg
                  className="Icon-Svg"
                  viewBox="0 0 12 12"
                >
                  <path
                    d="M6 8.5L10 4H2l4 4.5z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </div>
      <span
        className="Icon Icon_size_xs Icon_view_secondary IconTrash StepEditor__RemoveBtn"
        onClick={[Function]}
      >
        <svg
          className="Icon-Svg"
          viewBox="0 0 12 12"
        >
          <path
            clipRule="evenodd"
            d="M4 2a1 1 0 011-1h2a1 1 0 011 1h3v1h-1v7a1 1 0 01-1 1H3a1 1 0 01-1-1V3H1V2h3zm1 2H4v5h1V4zm3 0H7v5h1V4z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <button
        className="Button Button_size_s Button_view_ghost Button_width_default Button_form_default Button_withIcon MixFocus"
        onClick={[Function]}
        type="button"
      >
        <span
          className="Icon Icon_size_xs IconAdd Button-Icon Button-Icon_position_left"
        >
          <svg
            className="Icon-Svg"
            viewBox="0 0 12 12"
          >
            <path
              d="M6.5 2.5h-1v3h-3v1h3v3h1v-3h3v-1h-3v-3z"
            />
          </svg>
        </span>
        <span
          className="Button-Label"
        >
          Добавить риск
        </span>
      </button>
    </div>
  </div>
  <div
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <p
      className="StepEditor__SectionLabel"
    >
      Параметры
    </p>
    <p
      className="StepEditor__SelectLabel"
    >
      Стоимость
    </p>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <div
        className="Select Select_size_s Select_form_default Select_view_default StepEditor__ParametersSelect"
      >
        <div
          aria-expanded={false}
          aria-haspopup="listbox"
          className="Select-Control"
          id="9d9d871b-9355-4515-85e1-70cfe405f8e5"
        >
          <div
            className="Select-ControlInner"
          >
            <div
              className="Select-ControlValueContainer"
            >
              <input
                className="Select-FakeField"
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                readOnly={true}
                type="button"
              />
              <span
                className="Select-ControlValue"
                title="10 000 000 рублей"
              >
                10 000 000 рублей
              </span>
            </div>
          </div>
          <span
            className="Select-Indicators"
          >
            <button
              className="Select-IndicatorsDropdown"
              onClick={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                className="Icon Icon_size_xs IconSelect Select-DropdownIndicatorIcon"
              >
                <svg
                  className="Icon-Svg"
                  viewBox="0 0 12 12"
                >
                  <path
                    d="M6 8.5L10 4H2l4 4.5z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </div>
    </div>
    <div
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
    >
      <button
        className="Button Button_size_s Button_view_ghost Button_width_default Button_form_default Button_withIcon MixFocus"
        onClick={[Function]}
        type="button"
      >
        <span
          className="Icon Icon_size_xs IconAdd Button-Icon Button-Icon_position_left"
        >
          <svg
            className="Icon-Svg"
            viewBox="0 0 12 12"
          >
            <path
              d="M6.5 2.5h-1v3h-3v1h3v3h1v-3h3v-1h-3v-3z"
            />
          </svg>
        </span>
        <span
          className="Button-Label"
        >
          Добавить параметр
        </span>
      </button>
    </div>
  </div>
</div>
`;
